mindmeld/components/entity_resolver.py

```diff
@@ -562,38 +1151,575 @@ def _predict_exact_match(self, entity):
+            cnames2synonyms = {}
+            for syn, cnames in entity_mapping_synonyms.items():
+                for cname in cnames:
+                    items = cnames2synonyms.get(cname, [])
+                    items.append(syn)
+                    cnames2synonyms[cname] = items
+            dummy_new_synonyms_to_encode, dummy_new_synonyms_encodings = [], []
+            # assert dummy synonyms
+            for cname, syns in cnames2synonyms.items():
+                dummy_synonym = f"{cname} - SYNONYMS AVERAGE"
+                # update synonyms map 'cause such synonyms don't actually exist in mapping.json file
+                dummy_synonym_mappings = entity_mapping_synonyms.get(dummy_synonym, [])
+                dummy_synonym_mappings.append(cname)
+                entity_mapping_synonyms[dummy_synonym] = dummy_synonym_mappings
+                # check if needs to be encoded
+                if dummy_synonym in synonyms:
+                    continue
+                # if required, obtain dummy encoding and update collections
+                dummy_encoding = scipy.sparse.csr_matrix(
+                    np.max([synonyms_embs[synonyms[syn]].toarray() for syn in syns], axis=0)
+                )
+                dummy_new_synonyms_to_encode.append(dummy_synonym)
+                dummy_new_synonyms_encodings.append(dummy_encoding)
+            if dummy_new_synonyms_encodings:
+                dummy_new_synonyms_encodings = scipy.sparse.vstack(dummy_new_synonyms_encodings)
+            if dummy_new_synonyms_to_encode:
+                synonyms_embs = (
+                    dummy_new_synonyms_encodings if not synonyms else scipy.sparse.vstack(
+                        [synonyms_embs, dummy_new_synonyms_encodings])
+                )
+                synonyms.update(
+                    OrderedDict(zip(
+                        dummy_new_synonyms_to_encode,
+                        np.arange(len(synonyms), len(synonyms) + len(dummy_new_synonyms_to_encode)))
+                    )
+                )
+
+        # returns a sparse matrix
+        self._unique_synonyms = [*synonyms.keys()]
+        self._syn_tfidf_matrix = synonyms_embs
+
+    def _predict(self, nbest_entities, top_n):
+        """Predicts the resolved value(s) for the given entity using cosine similarity.
+
+        Args:
+            nbest_entities (tuple): List of one entity object found in an input query, or a list  \
+                of n-best entity objects.
+
+        Returns:
+            (list): The resolved values for the provided entity.
+        """
+
+        # encode input entity
+        # TODO: Use all provided entities (i.e all nbest_entities) like elastic search
+        top_entity = nbest_entities[0]  # top_entity
+        top_entity_vector = self._vectorizer.transform([top_entity.text])
+
+        similarity_scores = self._syn_tfidf_matrix.dot(top_entity_vector.T).toarray().reshape(-1)
+        similarity_scores = np.around(similarity_scores, decimals=4)
```

Comment:
Why do we have to round these when we are still sorting and returning ordinal values?

Reply:
Rounding is actually helping to bring the correct answers on to the top scored synonyms in case of BERT-models. Often times, if not rounded, the scores of top synonym (a wrong answer) and a right answer are so close with differences after nth decimal.

Reply:
Understood, could you add a small comment here on it?